sycl/doc/DeviceGlobal.md

```diff
@@ -0,0 +1,602 @@
+}
+
+} // namepsace (unnamed)
+} // namespace sycl::detail
+```
+
+Note that a SYCL application can legally call SYCL APIs even before `main()` by
+calling them from a global constructor.  However, the integration headers have
+been designed to ensure that the address of each device global variable is
+registered with the DPC++ runtime before the user's application could legally
+use the variable, even if that use occurs before `main()` executes.
+
+The user's application cannot legally use a device global variable until the
+variable's constructor has been called, otherwise the application would be
+using an unconstructed object which has undefined behavior by C++ rules.  Since
+all device globals must be defined at namespace scope, the C++ rules for the
+order of global constructors only guarantee that the device global will be
+constructed before subsequent global variables in the same translation unit.
+Therefore, a user application could reference a device global from another
+global constructor only if that global constructor is for an object defined
+*after* the device global in the same translation unit.  However, the
+integration header defines `__sycl_device_global_registerer` *before* all
+device globals in the user's translation unit.  Therefore, the address of all
+device global variables in the translation unit will be registered with the
+DPC++ runtime before any user code could legally use them.
+
+#### Handling shadowed device global variables
+
+The example above shows a simple case where the user's device global variables
+can all be uniquely referenced via fully qualified lookup (e.g.
+`::inner::Bar`).  However, it is possible for users to construct applications
+where this is not the case, for example:
+
+```
+sycl::device_global<int> FuBar;
+namespace {
+  sycl::device_global<int> FuBar;
+}
+```
+
+In this example, the `FuBar` variable in the global namespace shadows a
+variable with the same name in the unnamed namespace.  The integration footer
+can reference the variable in the global namespace as `::FuBar`, but there is
+no way to reference the variable in the unnamed namespace using fully qualified
+lookup.
+
+Such programs are still legal, though.  The integration footer can support
+cases like this by defining a temporary variable that holds the address of the
+shadowed device global:
+
+```
+namespace {
+const void *__sycl_UNIQUE_STRING = &FuBar;  // References 'FuBar' in the
+                                            // unnamed namespace
+}
+
+namespace sycl::detail {
+namespace {
+
+__sycl_device_global_registration::__sycl_device_global_registration() noexcept {
+  device_global_map::add(&::FuBar,
+    /* same string returned from __builtin_sycl_unique_stable_id(::FuBar) */);
+  device_global_map::add(::__sycl_UNIQUE_STRING,
+    /* same string returned from __builtin_sycl_unique_stable_id(::(unnamed)::FuBar) */);
+}
+
+} // namepsace (unnamed)
```

Comment:
namespace*

Reply:
Thanks.  Fixed in 6a21bb27b6d4a159dbabb043fdd29a5a338c1289.